_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,304 @@
+---
+layout: post
+title: "Kafka transactions (exactly-once semantics)"
+date: 2023-04-14
+author: federico_valeri
+---
+
+Apache Kafka is a distributed streaming platform that allows efficient and reliable processing of messages.
+It is widely used in various industries for data streaming applications, such as processing real-time data, event sourcing, and microservices integration.
+One of the key features of Kafka is the support for transactions, which provides exactly-once semantics (EOS) and is available since Kafka 0.11 ([KIP-98](https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging)).
+The support for EOS was recently extended to source connectors in Kafka Connect, which enables configuration-driven and fully transactional streaming pipelines ([KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors)).
+
+As messages flow through Kafka, they can be processed by multiple consumers, and ensuring that all consumers that should receive a message actually receive it once and only once can be crucial.
+The EOS guarantees that any set of operations inside Kafka will be completed atomically, either all of the operations succeed or none of them do.
```

Comment:
I agree, but this is only the intro and I'm clarifying this point in the considerations. There, I'll replace "external systems" with "processors with externally observable side-effects", because simply accessing an external system does not mean changing its state. Thanks.